### PR TITLE
Support RSA-PSS signature mechanisms using RSA-RAW

### PR DESCRIPTION
**Pull request by Jakuje**
This series of commits is based on https://github.com/NWilson/OpenSC/tree/pss-sigs (work from @NWilson), rebased on current master, resolved conflicts, fixed up the issues with corned cases, extended the support for different MGF1's independent of the hash algorithm and various corner cases.

I tested the code with PIV Test cards and Yubkey 4 with various mechanisms to the success. The test basic test cases are updated in wiki:

https://github.com/OpenSC/OpenSC/wiki/Using-pkcs11-tool-and-OpenSSL#sha1-rsa-pkcs-pss

I would be glad for feedback from @CardContact making sure that I did not break your use case with a cards having the direct PSS support.

Additionally, I extended also pkcs11-tool with an option to verify existing signatures, since the `C_Verify()` for PSS was already somehow implemented in the tool and might be useful for somebody.

What is not implemented is the possibility to configure the salt length of the PSS signatures. This would require either another bits in the flags bitarray or rework of the flags handling around the PSS padding creation. This could also use the MGF1 parameters that are at the moment accommodated in the same flags bitarray.

This is mostly first version of "something that works for me" and something for a first rounds of reviews about the design decisions, which might be adjusted at this moment.

Resolves: #529

Just one output of the sign&verification round:
```
[jjelen@t470s OpenSC (pss-sigs)]$ ./src/tools/pkcs11-tool -s --input-file data --output-file data.sig --module ./src/pkcs11/.libs/opensc-pkcs11.so --id 02 -m SHA256-RSA-PKCS-PSS
Using slot 0 with a present token (0x0)
Logging in to "SSH key".
Please enter User PIN: 
Using signature algorithm SHA256-RSA-PKCS-PSS
PSS parameters: hashAlg=SHA256, mgf=MGF1-SHA256, salt_len=32 B
Logging in to "SSH key".
Please enter context specific PIN: 
[jjelen@t470s OpenSC (pss-sigs)]$  ./src/tools/pkcs11-tool --verify --input-file data --signature-file data.sig --module ./src/pkcs11/.libs/opensc-pkcs11.so  --id 02 -m SHA256-RSA-PKCS-PSS
Using slot 0 with a present token (0x0)
Using signature algorithm SHA256-RSA-PKCS-PSS
PSS parameters: hashAlg=SHA256, mgf=MGF1-SHA256, salt_len=32 B
Signature is valid
```

##### Checklist

- [X] Documentation is added or updated
- [X] New files have a LGPL 2.1 license statement
- [X] PKCS#11 module is tested
- [ ] Windows minidriver is tested
- [ ] macOS tokend is tested

### CODE REVIEW

**Comment by mouse07410** [comment]
An extremely naive question. I thought we added RSA-PSS and RSA-OAEP to `pkcs11-tool`? And it is supported by OpenSSL via libp11. So is this PR supposed to replace that work? Complement it? Or...?

**Comment by Jakuje** [comment]
The previous work we done in #1169 and #1146 was limited to the `pkcs11-tool` and `pkcs11-spy` tools that can work with different PKCS#11 modules. The #1404 added support for on-card PSS padding in SC-HSM directly to OpenSC PKCS#11 module. This adds the support for PSS padding in OpenSC PKCS#11 module for all cards that support raw RSA operations by doing the PSS padding in software. Therefore, this work does not replace the previous work, but continues or complements it on another layer.

The libp11 support for PSS is also complement for all this work, since it allows using the PSS mechanisms from other applications using libp11.

> Also, the subject says "...using RSA-RAW". Does it mean it won't work with tokens/devices that support RSA-PSS natively (like YubiHSM2)?

Again, the yubihsm driver is not part of OpenSC so this change does not affect it at all. But we already have sc-hsm devices having native support for RSA-PSS (#1404) and these should not be affected by this change either, since they should provide the respective capability flags.

File: src/tools/pkcs11-tool.c [hunk 1]
```diff
@@ -1662,26 +1692,13 @@ static unsigned long figure_pss_salt_length(const int hash) {
...
-	if (!opt_mechanism_used)
-		if (!find_mechanism(slot, CKF_SIGN|CKF_HW, NULL, 0, &opt_mechanism))
-			util_fatal("Sign mechanism not supported");
+	unsigned long hashlen = 0;
 
-	fprintf(stderr, "Using signature algorithm %s\n", p11_mechanism_to_name(opt_mechanism));
-	memset(&mech, 0, sizeof(mech));
-	mech.mechanism = opt_mechanism;
-	pss_params.hashAlg = 0;
+	pss_params->hashAlg = 0;
```
Comment (frankmorgner) [inline src/tools/pkcs11-tool.c]: need to check for `pss_params!=NULL`?

**Comment by CardContact** [comment]
Why don't we use RSA_padding_add_PKCS1_PSS_mgf1() from OpenSSL rather then re-implementing PSS padding ?

**Comment by Jakuje** [comment]
> Would it be possible to add some testcases to p11test?

It is already there:
https://github.com/OpenSC/OpenSC/blob/master/src/tests/p11test/p11test_case_pss_oaep.c

> Why don't we use RSA_padding_add_PKCS1_PSS_mgf1() from OpenSSL rather then re-implementing PSS padding ?

Because this call requires complete RSA key object as an argument (to get just the modulus length), which we do not have in the place where we call the add-padding. I was considering using at least the `PKCS1_MGF1()`, but it was very cumbersome so I gave up.

File: src/libopensc/padding.c [hunk 2]
```diff
@@ -231,22 +237,200 @@ int sc_pkcs1_strip_digest_info_prefix(unsigned int *algorithm,
...
+	 * verified to match one of the above. If not, we will fail later
+	 */
+	return SC_ALGORITHM_RSA_HASH_NONE;
+}
+#endif
+
 /* general PKCS#1 encoding function */
 int sc_pkcs1_encode(sc_context_t *ctx, unsigned long flags,
-	const u8 *in, size_t in_len, u8 *out, size_t *out_len, size_t mod_len)
+	const u8 *in, size_t in_len, u8 *out, size_t *out_len, size_t mod_bits)
```
[the remark this answers is not in the document]
Comment (Jakuje) [reply src/libopensc/padding.c]: True. Fixed in other cases, but I am not sure what to do with the minidriver -- it is passing the last argument of 0, which is not valid argument in `sc_pkcs1_add_01_padding()`. Is this code path ever executed?

File: src/pkcs11/framework-pkcs15.c [hunk 3]
```diff
@@ -3780,35 +3772,80 @@ pkcs15_prkey_sign(struct sc_pkcs11_session *session, void *obj,
...
+		flags = SC_ALGORITHM_RSA_PAD_ISO9796;
+		break;
 	case CKM_RSA_PKCS_PSS:
-		rv = pkcs15_prkey_check_pss_param(pMechanism, (int)ulDataLen);
+		flags = SC_ALGORITHM_RSA_PAD_PSS;
+		/* The hash was done ouside of the module */
+		flags |= SC_ALGORITHM_RSA_HASH_NONE;
+		/* Omited parameter can use MGF1-SHA1 ? */
+		if (pMechanism->pParameter == NULL) {
+			flags |= SC_ALGORITHM_MGF1_SHA1;
```
[the remark this answers is not in the document]
Comment (Jakuje) [reply src/pkcs11/framework-pkcs15.c]: If I read it right, it does not say anything about default so in strictly following the PKCS#11, it should require all the parameters always.
AFAIK, this was in the original implementation with a good will of setting some default as most of the other applications.

File: src/pkcs11/framework-pkcs15.c [hunk 4]
```diff
@@ -5069,11 +5192,8 @@ register_mechanisms(struct sc_pkcs11_card *p11card)
 #endif /* ENABLE_OPENSSL */
 	}
 
-	/* TODO support other padding mechanisms */
-
 	if (rsa_flags & SC_ALGORITHM_RSA_PAD_PSS) {
-		mech_info.flags &= ~(CKF_DECRYPT|CKF_VERIFY);
-
+		mech_info.flags &= ~(CKF_DECRYPT|CKF_ENCRYPT);
```
[the remark this answers is not in the document]
Comment (Jakuje) [reply src/pkcs11/framework-pkcs15.c]: Yes. The PSS mechanism can be used for verification (using OpenSSL), but not at all for encryption.

File: src/pkcs11/openssl.c [hunk 5]
```diff
@@ -477,13 +496,95 @@ CK_RV sc_pkcs11_verify_data(const unsigned char *pubkey, int pubkey_len,
...
+				unsigned int tmp_len;
+
+				EVP_DigestFinal(md_ctx, tmp, &tmp_len);
+				data = tmp;
+				data_len = tmp_len;
+			}
+			rv = CKR_SIGNATURE_INVALID;
+			if (data_len == EVP_MD_size(pss_md)
+			    && RSA_verify_PKCS1_PSS_mgf1(rsa, data, pss_md,
+			    mgf_md, rsa_out, EVP_MD_size(pss_md)/*sLen*/) == 1)
```
[the remark this answers is not in the document]
Comment (Jakuje) [reply src/pkcs11/openssl.c]: Well ... that is a question what is a good indent. unfortunately, there is no coding style guide in OpenSC. This one is OpenBSD/Kernel source file style guide (KNF) which I got used to recently:

>  Indentation is an 8 character tab. Second level indents are four spaces. All code should fit in 80 columns. 

https://man.openbsd.org/style

File: src/tools/pkcs11-tool.c (hunk 1 above)
Comment (Jakuje) [reply src/tools/pkcs11-tool.c]: Yes. Added.

File: src/pkcs11/framework-pkcs15.c [hunk 6]
```diff
@@ -3780,35 +3772,80 @@ pkcs15_prkey_sign(struct sc_pkcs11_session *session, void *obj,
...
+		break;
+	case CKM_RSA_X_509:
+		flags = SC_ALGORITHM_RSA_RAW;
+		break;
+	case CKM_RSA_X9_31:
+		flags = SC_ALGORITHM_RSA_PAD_ANSI;
+		break;
+	case CKM_RSA_9796:
+		flags = SC_ALGORITHM_RSA_PAD_ISO9796;
+		break;
```
[the remark this answers is not in the document]
Comment (Jakuje) [reply src/pkcs11/framework-pkcs15.c]: If I remember well, I tried this mechanism manually and it worked. But automated test would make sense. I will have a look into that.

[the remark this answers is not in the document]
Comment (Jakuje) [reply src/pkcs11/framework-pkcs15.c]: See the other comment -- I decided to remove this part since there is no good use of that.

File: src/pkcs11/openssl.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (frankmorgner) [reply src/pkcs11/openssl.c]: - You're mixing tabs and spaces
- Opening the bracket with `RSA_verify_PKCS1_PSS_mgf1(` should increment the the indent by one level

[the remark this answers is not in the document]
Comment (Jakuje) [reply src/pkcs11/openssl.c]: 1) Yes, this is how the above referenced style does it. If you prefer other style, lets create some formal contribution guide, declaring the preference.
2) That is a good point. Fixed.

File: src/libopensc/padding.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (frankmorgner) [reply src/libopensc/padding.c]: The minidriver only sets hash flags, no padding flags. Hence, the `mod_len` argument is not used in `sc_pkcs1_encode()`/`sc_pkcs1_add_01_padding()`. So regarding your PR, you could ignore the use of `sc_pkcs1_encode()` in minidriver.c.

However, I'm not sure if the minidriver implementation is optimal. Because after calling `sc_pkcs1_encode()` with the flag for hashing:
https://github.com/OpenSC/OpenSC/blob/3cc2670f3e5f30d0d244b2d07d9f959f3e840af7/src/minidriver/minidriver.c#L4742-L4762
an additional call to sc_pkcs15_compute_signature is made
https://github.com/OpenSC/OpenSC/blob/3cc2670f3e5f30d0d244b2d07d9f959f3e840af7/src/minidriver/minidriver.c#L4823-L4830
the second time, only the padding flag is used. I think the same could have been achieved by just calling `sc_pkcs15_compute_signature()` with both, padding and hashing flags. `sc_pkcs15_compute_signature()` will then call `sc_pkcs1_encode()` if needed. @vletoux, what do you think?

**Comment by vletoux** [comment]
Quoting the mini driver source code, PSS is marked as unsupported

if (CARD_PADDING_PSS == pInfo->dwPaddingType)   {
			logprintf(pCardData, 0, "unsupported paddingtype CARD_PADDING_PSS\n");
			dwret = SCARD_E_UNSUPPORTED_FEATURE;
			goto err;
		}

Some additional work has to be done on the mini driver side.
Vincent

**Comment by frankmorgner** [comment]
Thanks for the pointer!

What do you think about directly calling `sc_pkcs15_compute_signature()` without a prior call to `sc_pkcs1_encode()`?

**Comment by Jakuje** [comment]
Thank you. From what I read in the comments, there might or might not be needed to address some of the following issues:
 * minidriver - unfortunately, I can not help in this way
 * Change of the constant values for `SC_ALGORITHM_*` values -- I hope nobody used them as their value data, but even if so, we will probably not find earlier than something will break.
 * Default values for PSS parameters -- we benevolently accept some of the missing parameters, but we probably should not do that to follow the PKCS#11 specification

I squashed the changes and rebased them on master, leaving the less related in separate commits.

**Comment by vletoux** [comment]
About calling sc_pkcs15_compute_signature directly, I remind that sc_pkcs1_encode was used to append the OID of the algorithm to the hash before signing it.
(difference between "normal signature" = oid + hash and "raw signature" = hash only)
If sc_pkcs15_compute_signature is doing the same (adding the oid before the hash) with a flag, I see no reason to add this step.
But the minidriver has to be change to reflect the change and be sure the oid is added before doing the signature.
